src/opentime/rationalTime.h

```diff
@@ -16,46 +16,84 @@ enum IsDropFrameRate : int {
     
 class RationalTime {
 public:
-    explicit RationalTime(double value = 0, double rate = 1)
-    : _value {value}, _rate {rate} {}
+    explicit RationalTime(double value = 0, double rate = 1) : _value {value}, _rate {rate} {}
+    RationalTime(RationalTime && rh) { _value = rh._value; _rate = rh._rate; }
+    RationalTime(RationalTime const& rh) { _value = rh._value; _rate = rh._rate; }
+    RationalTime& operator= (RationalTime const& rh) { _value = rh._value; _rate = rh._rate; return *this; }
+    RationalTime& operator= (RationalTime && rh) { _value = rh._value; _rate = rh._rate; return *this; }
+    ~RationalTime() = default;
+
+    bool is_invalid_time() const { return _rate <= 0; }
+    double value() const { return _value; }
+    double rate() const { return _rate; }
+    
+    void set_value(double v) { _value = v; }
+    void set_rate(double r) { _rate = r; }
+    
+    RationalTime floor() const {
+        return RationalTime { std::floor(_value), _rate };
+    }
+
+    RationalTime floor(double rate) const {
+        return rescaled_to(rate).floor();
+    }
+    
+    RationalTime ceil() const {
+        return RationalTime { std::ceil(_value), _rate };
+    }
 
-    RationalTime(RationalTime const&) = default;
-    RationalTime& operator= (RationalTime const&) = default;
+    RationalTime ceil(double rate) const {
+        return rescaled_to(rate).ceil();
+    }
 
-    bool is_invalid_time() const {
-        return _rate <= 0;
+    RationalTime round_nearest() const {
+        return RationalTime(_value + 0.5 / _rate, _rate).floor();
     }
     
-    double value() const {
-        return _value;
+    RationalTime round_nearest(double rate) const {
+        return rescaled_to(rate).round_nearest();
+    }
```

Comment:
The code I am working with needs ceil/floor snapping to frame all the time in hundreds of places. I would rather include them than force equally error prone shenanigans like making someone code t.floor(t.rate). In the code I'm working with there had been some bad assumptions about 24, and smoking out things like t.floor(24), or t.floor(t2) consumed enough time, that I think having the frame floor and ceil is a definitely a win. 

I see your point though, the ambiguous naming is not a win. How about a different name?

```cpp
RationalTime rate_floor() const;    // frame rate natural floor
RationalTime rate_floor(double) const; // arbitrary rate floor
RationalTime rate_ceil() const; // frame rate natural ceil
RationalTime rate_ceil() const; // arbitrary rate ceil
```

Reply:
Upon reflection I see your point. I also think we may lack some clarity in the API about what operations should be considered "safe" for preserving rate vs. which are "unsafe" - not to mention the definition of "safe" may depend on the use domain.

 There are probably cases where a programmer can be reasonably confident that the `RationalTime` values they are using have consistent `rate` values and it's not the APIs job to protect them against when that assumption is bad. This is also more true for the C++ api where there is less expectation of guardrails.

To be further pedantic, I can see how `rate_floor` and `rate_ceil` can read as "floor to rate" or "ceil to rate", but on first reading I thought "floor the rate".
Would it be more clear to call them `value_floor` and `value_ceil`? This way it's clear the `value` component is being operated on and the optional argument is about explicitly providing the quantization? This would also be similar to how the python `round` function works where it rounds to whole numbers (an implicit quantization) by default, but you can specify an `ndigits` arg to provide an explicit quantization level.

Reply:
The operation we are creating is "snap to the smallest frame not less than t" --- ceil, and "snap to the largest frame not greater than t" --- floor. That's how the c standard defines them (smallest integer not less than val, greatest integer not greater than t)

Think think think,,,, I think I like your value_floor/ceil suggestion as it describes exactly the math that will occur

Reply:
I haven't thought of an API change that makes things more clear, as opposed to simply moving things about. My preference at this point is to leave it as is, absent further proposals.